### PR TITLE
Global browser setting / Private mode option

### PR DESCRIPTION
**Pull request by onesounds**
**Plan**
- It integrates browser settings of all web-related plugins into default settings.
- If necessary, make private mode available.
  - Currently, there are only two on/off configurations, but it would be nice if factors could be added and used only when necessary. (e.g google.com-p)

image

### CODE REVIEW

**Comment by jjw24** [comment]
could we resolve conflict on this one please @taooceros @onesounds

**Comment by onesounds** [comment]
This pr has not been completed and old. It was created only to create a setting window and pass the work to @taooceros. I thought it would be done quickly because it was a similar type of work to the file manager, but I close it first.  taooceros is so busy.

**Comment by taooceros** [comment]
Lol don't close it. I can do it today🤣

**Comment by onesounds** [comment]
oh, is it possible? I don't know it's good UI. because other programs has no that option. they actually works based on OS setting. We actually only need private mode or new window/new tab.

**Comment by taooceros** [comment]
> oh, is it possible? I don't know it's good UI. because other programs has no that option. they actually works based on OS setting. We actually only need private mode or new window/new tab.

I actually thinking about something like per profile browsering, but that may be a bit overkill?

**Comment by onesounds** [comment]
> I actually thinking about something like per profile browsering, but that may be a bit overkill?

It seems that you need a profile to use private mode easily. If there's no other opinion, let's just do this.

**Comment by jjw24** [comment]
cant seem to click the done button

image

**Review by jjw24 — COMMENTED** [review]
cant seem to click the done button

image

**Comment by jjw24** [comment]
the tab/new window option switches around everytime the default browser window is opened

File: Flow.Launcher.Infrastructure/UserSettings/CustomBrowserViewModel.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+            return new CustomBrowserViewModel
+            {
+                Name = Name,
+                Path = Path,
+                PrivateArg = PrivateArg,
+                EnablePrivate = EnablePrivate,
+                Editable = Editable
+            };
+        }
+
```
Comment (jjw24) [inline Flow.Launcher.Infrastructure/UserSettings/CustomBrowserViewModel.cs:26]:
```suggestion
```

File: Flow.Launcher.Infrastructure/UserSettings/CustomBrowserViewModel.cs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    }
+
```
Comment (jjw24) [inline Flow.Launcher.Infrastructure/UserSettings/CustomBrowserViewModel.cs:28]:
```suggestion
```

File: Flow.Launcher.Plugin/Interfaces/IPublicAPI.cs [hunk 3]
```diff
@@ -212,5 +212,8 @@ public interface IPublicAPI
         /// <param name="DirectoryPath">Directory Path to open</param>
         /// <param name="FileName">Extra FileName Info</param>
         public void OpenDirectory(string DirectoryPath, string FileName = null);
+
+
```
Comment (jjw24) [inline Flow.Launcher.Plugin/Interfaces/IPublicAPI.cs:216]:
```suggestion
```

File: Flow.Launcher.Infrastructure/UserSettings/CustomBrowserViewModel.cs [hunk 4]
```diff
@@ -0,0 +1,32 @@
+using Flow.Launcher.Plugin;
+
+namespace Flow.Launcher.Infrastructure.UserSettings
+{
+    public class CustomBrowserViewModel : BaseModel
```
Comment (jjw24) [inline Flow.Launcher.Infrastructure/UserSettings/CustomBrowserViewModel.cs:5]: why not put this  in the same location as the other vms?

File: Flow.Launcher/PublicAPIInstance.cs [hunk 5]
```diff
@@ -209,6 +209,23 @@ public void OpenDirectory(string DirectoryPath, string FileName = null)
             explorer.Start();
         }
 
+        public void OpenUrl(string url)
+        {
+            using var process = new Process();
```
Comment (jjw24) [inline Flow.Launcher/PublicAPIInstance.cs:214]:
```suggestion
```

File: Flow.Launcher/SelectBrowserWindow.xaml.cs [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+using System.Windows.Input;
+using System.Windows.Media;
+using System.Windows.Media.Imaging;
+using System.Windows.Shapes;
+
+namespace Flow.Launcher
+{
+    /// <summary>
+    /// SelectBrowserWindow.xaml에 대한 상호 작용 논리
+    /// </summary>
```
Comment (jjw24) [inline Flow.Launcher/SelectBrowserWindow.xaml.cs:23]:
```suggestion
```

File: Flow.Launcher/SettingWindow.xaml.cs [hunk 7]
```diff
@@ -120,6 +120,12 @@ private void OnSelectFileManagerClick(object sender, RoutedEventArgs e)
             fileManagerChangeWindow.ShowDialog();
         }
 
+        private void OnSelectDefaultBrowserClick(object sender, RoutedEventArgs e)
+        {
+            SelectBrowserWindow test = new SelectBrowserWindow(settings);
+            test.ShowDialog();
```
Comment (jjw24) [inline Flow.Launcher/SettingWindow.xaml.cs:126]:
```suggestion
            var browserWindow = new SelectBrowserWindow(settings);
            browserWindow.ShowDialog();
```

File: Plugins/Flow.Launcher.Plugin.BrowserBookmark/Views/CustomBrowserSetting.xaml [hunk 8]
```diff
@@ -1,38 +1,192 @@
...
+                Name="btnConfirm"
+                Margin="15"
+                Click="ConfirmCancelEditCustomBrowser"
+                Content="Confirm" />
+            <Button
+                Margin="15"
+                Click="ConfirmCancelEditCustomBrowser"
+                Content="Cancel" />
         </StackPanel>
+        -->
```
Comment (jjw24) [inline Plugins/Flow.Launcher.Plugin.BrowserBookmark/Views/CustomBrowserSetting.xaml:190]: @onesounds is there a reason we are keep this commented out code?

File: Plugins/Flow.Launcher.Plugin.WebSearch/SettingsControl.xaml.cs [hunk 9]
```diff
@@ -21,9 +21,7 @@ public SettingsControl(PluginInitContext context, SettingsViewModel viewModel)
             _context = context;
             _settings = viewModel.Settings;
             DataContext = viewModel;
-            browserPathBox.Text = _settings.BrowserPath;
-            NewWindowBrowser.IsChecked = _settings.OpenInNewBrowser;
-            NewTabInBrowser.IsChecked = !_settings.OpenInNewBrowser;
+
```
Comment (jjw24) [inline Plugins/Flow.Launcher.Plugin.WebSearch/SettingsControl.xaml.cs:24]:
```suggestion
```

**Review by jjw24 — CHANGES_REQUESTED** [review]

File: Flow.Launcher.Infrastructure/UserSettings/CustomBrowserViewModel.cs (hunk 4 above)
Comment (taooceros) [reply Flow.Launcher.Infrastructure/UserSettings/CustomBrowserViewModel.cs:5]: because it needs to be accessed in Setting

**Comment by jjw24** [comment]
- The new tab/window toggle is still switching around when I open the browser setting

- private mode doesn't seem to be working

- @taooceros  for some reason the GitHub plugin fails to open the repo via url ('pressing ctrl + enter on the repo result')

File: Plugins/Flow.Launcher.Plugin.BrowserBookmark/Views/CustomBrowserSetting.xaml (hunk 8 above)
Comment (onesounds) [reply Plugins/Flow.Launcher.Plugin.BrowserBookmark/Views/CustomBrowserSetting.xaml:190]: I missed. removed.

**Comment by jjw24** [comment]
> private mode doesn't seem to be working

working now

**Comment by jjw24** [comment]
> @taooceros for some reason the GitHub plugin fails to open the repo via url ('pressing ctrl + enter on the repo result')

sorted

**Comment by jjw24** [comment]
remaining two:

1. url open with auto suggestion doesnt work
2. firefox new window with private wrong work (just open private window without url. )

**Comment by onesounds** [comment]
ok. firefox is fine now. Sometimes it looks like a problematic movement, but it is a situation created by Firefox. It's not about us.
- https://wiki.mozilla.org/Firefox/CommandLineOptions

**Comment by onesounds** [comment]
url open with auto suggestion works. confirmed.

File: Flow.Launcher.Plugin/Interfaces/IPublicAPI.cs [hunk 10]
```diff
@@ -212,5 +212,10 @@ public interface IPublicAPI
         /// <param name="DirectoryPath">Directory Path to open</param>
         /// <param name="FileName">Extra FileName Info</param>
         public void OpenDirectory(string DirectoryPath, string FileName = null);
+
+        /// <summary>
+        /// Open Url in the configured default browser for Flow's Settings.
```
Comment (jjw24) [inline Flow.Launcher.Plugin/Interfaces/IPublicAPI.cs:217]:
```suggestion
        /// Opens the url. The browser and mode used is based on what's configured in Flow's default browser settings.
```

**Review by jjw24 — APPROVED** [review]
great work everyone!

**Pull request merged**